keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,722 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes currently does not support the use of [swap memory](https://en.wikipedia.org/wiki/Paging#Linux) on Linux, as it is difficult to provide guarantees and account for pod memory utilization when swap is involved. As part of Kubernetes’ earlier design, [swap support was considered out of scope](https://github.com/kubernetes/kubernetes/issues/7294).
+
+However, there are a [number of use cases](#user-stories) that would benefit from Kubernetes nodes supporting swap. Hence, this proposal aims to add swap support to nodes in a controlled, predictable manner so that Kubernetes users can perform testing and provide data to continue building cluster capabilities on top of swap.
+
+## Motivation
+
+There are two distinct types of user for swap, who may overlap:
+- node administrators, who may want swap available for node-level performance tuning and stability/reducing noisy neighbour issues
+- application developers, who have written applications that would benefit from using swap memory
+
+There are hence a number of possible ways that one could envision swap use on a node.
+
+### Scenarios
+
+1. Swap is enabled on a node's host system, but the CRI does not permit Kubernetes workloads to use swap. (This scenario is a prerequisite for the following use cases.)
+1. Swap is enabled at the node level. The CRI can be globally configured to permit user workloads scheduled on the node to use some quantity of swap.
+1. Swap is set on a per-workload basis. The CRI sets permitted swap utilization on each individual workload.
+
+This KEP will be limited in scope to the first two scenarios. The third can be addressed in a follow-up KEP. The enablement work that is in scope for this KEP will be necessary to implement the third scenario.
+
+
+### Goals
+
+- On Linux systems, when swap is provisioned and available, Kubelet can start up with swap on.
+- Configuration is available for CRI to set swap utilization available to Kubernetes workloads, defaulting to 0 swap.
+- Cluster administrators can enable and configure CRI swap utilization on a per-node basis.
+- Use of swap memory with both cgroupsv1 and cgroupsv2 is supported.
+
+### Non-Goals
+
+- Provisioning swap. Swap must already be available on the system.
+- Setting [swappiness]. This can already be set on a system-wide level outside of Kubernetes.
```

Comment:
This KEP doesn't enable setting per-workload swap utilization, so I think it's YAGNI on per-workload swappiness configuration in the CRI. When we're ready to add per-workload swap configuration to k8s, then we should add it IMO.

> but for system workloads (e.g those in /system.slice, i.e. kubelet, container runtime, etc), you don't want them to swap at all (and want a zero swappiness). 

Typically one wouldn't run the container runtime, kubelet, etc. using k8s though, they're prerequisites for a node that run directly on the host. So potentially this is still individually configurable for them, no?

Reply:
> When we're ready to add per-workload swap configuration to k8s, then we should add it IMO.

Make sense!

> So potentially this is still individually configurable for them, no?

Good point, I guess what you would do in that case is set `vm.swappiness` so that it applies system wide, but then you would have to edit all of your systemd units for kubelet, container runtime etc and override their swap settings. Might be helpful to note somewhere if that's the recommended path to configure swappiness since I would imagine most existing kubelet/container runtime units don't override swap settings.